### DISCUSSION TITLE
How can I get metadata from request

### DISCUSSION THREAD

**Discussion by rrolla** [savonet/liquidsoap#2870] [Q&A]
**I want to return which jingle will start playing over current track when executing command in telnet**
```ruby
jinglesOverSource = playlist(id="jingleOverSource", reload_mode="watch", "~/content/jingles")
playJingleOverRef = ref(false)

def jingleOver()
   if null.defined(jinglesOverSource.current()) then
       jingleOverRequest = jinglesOverSource.current()
       jingleOverMetadata = request.metadata(jingleOverRequest ?? request.create(null()))
       print("------ jingleOver LASTMETADATA jingleOverMetadata ----- #{jingleOverMetadata}")
   end
   log("jingleOver called")
   playJingleOverRef := true
end

def addRandomJingleCommand(~namespace, ~command)
    server.register(
        namespace=string_of(namespace),
        usage=string_of(command),
        description="Play random jingle over stream",
        command,
        fun(_) -> begin jingleOver() "Done!" end
    )
 end

addRandomJingleCommand(namespace="gradio.one", command="randomJingle")

```

But I got error
```ruby
Warning 3: This expression should have type unit.
At /home/liquidsoap/config/jingles/jingle-over.liq, line 8, char 81-87:
        jingleOverMetadata = request.metadata(jingleOverRequest ?? request.create(null()))
 
 Error 5: this value has type
   _?
 but it should be a subtype of
   string
```

On documentation, there are return types

playlist operator https://www.liquidsoap.info/doc-2.1.3/reference.html#playlist returns source 

```ruby
-> source(audio='a, video='b, midi='c)
```
and `source` have method `current` that returns nullable request

```ruby
current (of type () -> request?): Get the request currently being played.
```

so to access current request I need to call

```ruby
jingleOverRequest = jinglesOverSource.current()
```

and `request` have method `metadata` https://www.liquidsoap.info/doc-2.1.3/reference.html#request.metadata
so to get metadata I need to call
```ruby
jingleOverMetadata = request.metadata(jingleOverRequest ?? request.create(null()))
```

**Comment by vitoyucepi** [comment]
Hi @rrolla,
I assume this is what you want to do.
```ruby
jingleOverRequest = jinglesOverSource.current()
if null.defined(jinglesOverRequest) then
  jingleOverMetadata = request.metadata(null.get(jingleOverRequest))
end
```
Yet I'm heavily concerned about that snippet overall.
I think telnet is not the right way to interact with the liquidsoap if you can use HTTP.
Also, if the track is current, it's possible to call `last_metadata()`.

**Reply by rrolla** [reply]
@vitoyucepi with your tweak now liquidsoap starts but it never get in this condition
```ruby
if null.defined(jingleOverRequest) then
```

If I use `last_metadata()`it return previous jingle what was played, not current one that should play

As this command is rearly needed it will be perfectly fine to just do in telnet, instead play random jingle and then post to backend vai http call

**Reply by vitoyucepi** [reply]
`null.defined` should work as `jinglesOverRequest != null()`.
https://github.com/savonet/liquidsoap/blob/a49a9dfb6dffb88a6c0ddf6f817ec7b0eb0d4b31/libs/metadata.liq#L149-L154

It is the same as yours.
> ```ruby
> if null.defined(jinglesOverSource.current()) then
>      jingleOverRequest = jinglesOverSource.current()
>      jingleOverMetadata = request.metadata(jingleOverRequest ?? request.create(null()))
>      print("------ jingleOver LASTMETADATA jingleOverMetadata ----- #{jingleOverMetadata}")
>  end
> ```

**Reply by rrolla** [reply]
Looks like problem is with time, i want to get metadata before its requested, I added logs 
```ruby
    jingleOverRequest = jinglesOverSource.current()
    print("jinglesOverSource.current() #{jingleOverRequest}")
```

and most of the time it's null, only if I'm spamming telnet command then I can get request there

Basically i need same functionality whats already in telnet `jingleOverSource.next`

![image](https://user-images.githubusercontent.com/3932546/215345968-59449bd1-0454-4e48-b771-01e1304b235c.png)

if i execute this command i can get jingle that will play as next
```bash
jingleOverSource.next
[ready] annotate:type="jingles":/home/liquidsoap/content/jingles/Now_we_will_listen_to_another_great_tune__right_here_on_G-Radio_-4245754115.mp3
END

```

**Reply by vitoyucepi** [reply]
You can use `playlist.queue()` in combination with `playlist.fetch()`.
